**Admin**: [Getting started](admin/getting_started.md) * [Setting up an analysis](admin/setting_up_an_analysis_guidelines.md) * [Data Management](admin/data_management.md) * [Reproducible Research](admin/reproducible_research.md) * [Methods for manuscripts](admin/method_snippets.md) * [Downloading data](admin/download_data.md) * [Globus](admin/Globus.md) * [Acknowledging funding](admin/acknowledging_funding) *  [Initial consults](admin/initial_consults.md) * [Chargeback models](admin/chargeback_models.md) * [Archiving project folders](admin/archive_folders_to_standby.md) * [Consulting resources](admin/consulting_resources.md) * [Pull for knowledge work instead of push](https://dspace.mit.edu/bitstream/handle/1721.1/115364/Repenning_Pull%20for%20Knowledge%20Work_Full.pdf?sequence=1&isAllowed=y) * [Scripts for tracking storage on O2](admin/scripts_for_data_management_on_o2) 


**RC**: * [O2 OpenPortal](rc/O2_portal_errors.md) * [Connection to HPC](rc/connection-to-hpc.md) * [FASRC](admin/FASRC.md) * [Keep command alive](rc/keepalive.md) * [IPython notebook](rc/ipython-notebook-on-O2.md) * [Manage files](rc/manage-files.md) * [O2 tips](rc/O2-tips.md) * [Scheduler](rc/scheduler.md) * [Jupyter notebooks](rc/jupyter_notebooks.md) * [tmux tips](rc/tmux.md) * [FAS virtual desktops](rc/openondemand.md)*[Arrays in Slurm](rc/arrays_in_slurm.md)*[Common O2 portal Issues](rc/O2_portal_errors.md) 

**bcbio**: [docs](https://bcbio-nextgen.readthedocs.io/en/latest/) * [issues](https://github.com/bcbio/bcbio-nextgen/issues) * [validation](https://github.com/bcbio/bcbio_validations) * [running in parallel](https://bcbio-nextgen.readthedocs.io/en/latest/contents/parallel.html) * [genomes](bcbio/bcbio_genomes.md) * [tips](bcbio/bcbio_tips.md) * [RNA-seq workflow by Mary](bcbio/bcbio_workflow_mary.md) * [installation by Michael](https://steinbaugh.com/posts/install-bcbio.html) * [installation by Sergey](https://github.com/naumenko-sa/bioscripts/blob/master/bcbio/bcbio.upgrade.sh) * [RNA-seq workflow by Systems Pharmacology lab](https://labsyspharm.github.io/rnaseq/) * [gtf/gff validators](bcbio/gtf_gff_validator.md) * [Create hybrid mammal virus genome](bcbio/Creating_Hybrid_Mammal_Viral_Reference_Genome.md) * [python package in 2022](https://mathspp.com/blog/how-to-create-a-python-package-in-2022)

**Bulk RNA-seq**: [Count normalization methods](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html) * [Dexseq](rnaseq/dexseq.Rmd) * [Failure types](rnaseq/failure_types) * [IRFinder_report](rnaseq/IRFinder_report.md) * [RepEnrich2 guide](rnaseq/RepEnrich2_guide.md) * [IRFinder](rnaseq/running_IRFinder.md) * [Strandedness](rnaseq/strandedness.md) * [Tools](rnaseq/tools.md) * [Bibliography](rnaseq/bibliography.md) * [Bibliography.bib](rnaseq/bcbio_rnaseq.bib) * [ASE](rnaseq/ase.md) * [eQTL](http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/) * [GTEx](https://gtexportal.org/home/) * [SummarizedExperiment templates for QC and DE analyses](https://github.com/naumenko-sa/crt/blob/master/differential_expression/) * [bcbioLite](https://github.com/vbarrera/bcbioLite) * [Cluster profiler](https://yulab-smu.github.io/clusterProfiler-book/) * [ggsashimi](https://github.com/guigolab/ggsashimi) * [Differential exon usage with Suppa2](https://github.com/comprna/SUPPA) * [cell type deconvolution](https://www.synapse.org/#!Synapse:syn15589870/wiki/606797) * [webgestalt](http://www.webgestalt.org/) * [Differential splicing with Leafcutter](https://github.com/hbc/knowledgebase/blob/master/rnaseq/running_leafcutter.md)

**CHIP-seq**: [ENCODE Guidelines](http://genome.cshlp.org/cgi/pmidlookup?view=long&pmid=22955991) * [CHIp-seq intro](https://hbctraining.github.io/Intro-to-ChIPseq/schedule/2-day.html) * [Tools](chipseq/tools.md) * [Cut&RUN](chipseq/cutandrun.md) * [Antibody column on metadata](chipseq/metadata.md) * [Cut&RUN for tfBS](https://star-protocols.cell.com/protocols/944) * [Cut&Run tools2](https://github.com/fl-yu/CUT-RUNTools-2.0) * [TF regulons](https://bioconductor.org/packages/release/data/experiment/html/dorothea.html)

**Misc**: [Git tips](misc/git.md) * [Organized papers](https://github.com/hbc/knowledgebase/blob/master/misc/organized_papers.md) * [AWS](misc/aws.md) * [Core resources](misc/core_resources.md) * [General NGS](misc/general_ngs.md) * [Orphan improvements](misc/orphan_improvements.md) * [OSX](misc/OSX.md) * [Snakemake](misc/snakemake-example-pipeline) * [miRNA](misc/miRNA.md) * [detect sample mislabeling](https://github.com/brentp/somalier) * [Get SRA links](https://sra-explorer.info) * [GNU parallel](https://opensource.com/article/18/5/gnu-parallel) * [GEO Submissions](misc/GEO_submissions.md) * [Report FAQs Snippet](misc/FAQs.md) * [teaching materials from Liu lab](https://liulab-dfci.github.io/teaching) * [Connected papers](https://www.connectedpapers.com/) * [Speciation genomics tutorial](https://speciationgenomics.github.io/) * [Custom Genome with Reform](misc/Reform_python.md) * [Multiomics Factor Analysis](misc/multiomics_factor_analysis.md) * [Bioinformatics course](https://liulab-dfci.github.io/bioinfo-combio/) * [BARC SOP - WI MIT](http://barcwiki.wi.mit.edu/wiki/SOPs) * [Babraham courses](https://www.bioinformatics.babraham.ac.uk/training.html) *[Illumina Dragen](https://support-docs.illumina.com/SW/DRAGEN_v38/Content/SW/FrontPages/DRAGEN.htm) * [JTLeek group genomics papers](https://github.com/jtleek/genomicspapers) * [R-Metabolomics](https://rformassspectrometry.github.io/metaRbolomics-book/) * [Nextflow training](https://training.seqera.io/) * [IOWA State U bioinformatics core knowledgebase](https://bioinformaticsworkbook.org/#gsc.tab=0) * [Long-read presentation from Jihe](long_read_data) * [Creative science](https://night-science.org/a-course-on-the-creative-scientific-process/) * [bioinformatics tutorials NCI](https://bioinformatics.ccr.cancer.gov/btep/btep-video-archive-of-past-classes/) * [Bioinformarics workbook](https://bioinformaticsworkbook.org)

**Nanopore**: [Master of pores from CRG](https://github.com/biocorecrg/master_of_pores) * [Tombo](https://nanoporetech.github.io/tombo/) * [Megalodon](https://github.com/nanoporetech/megalodon) * [Processing Direct RNA Sequencing data on O2](https://github.com/hbc/knowledgebase/blob/master/long_read_data/nanopore_DRS_workflow.md)

**Multi-omics integration**: [Diablo paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6735831/) [vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/mixOmics/inst/doc/vignette.html#n-integration-06) * [totalVI - citeSeq](https://streetslab.berkeley.edu/tools/totalvi/)

**R**: [.Rprofile](r/.Rprofile) * [RShiny](r/rshiny_server.md) * [htmlwidgets](/r/htmlwidgets) * [Tip and tricks](r/R-tips-and-tricks.md) * [Complex heatmap](https://jokergoo.github.io/ComplexHeatmap-reference/book/) * [Fundamentals of data visualization](https://serialmentor.com/dataviz/) * [Big Book of R](https://www.bigbookofr.com/index.html) * [Statistical rethinking](https://bookdown.org/content/4857/) * [and other bookdown books](https://bookdown.org/) * [Boxplot evolution](https://www.cedricscherer.com/2019/05/17/the-evolution-of-a-ggplot-ep.-1/) * [Intro to stat computing](https://biodatascience.github.io/statcomp/) * [data to viz](https://www.data-to-viz.com/) * [modern statistics for modern biology](https://www.huber.embl.de//msmb/index.html) * [Heatmap examples](https://jokergoo.github.io/) * [Learniing statistics with R](https://learningstatisticswithr.com/) * [Understanding statistics and experimental design](https://library.oapen.org/bitstream/handle/20.500.12657/23029/1007132.pdf) * [data analysis and prediction algorithms with R](https://rafalab.github.io/dsbook/) * [R for data analysis](https://trevorfrench.github.io/R-for-Data-Analysis/) * [Exclude ranges](https://bioconductor.org/packages/release/data/annotation/vignettes/excluderanges/inst/doc/excluderanges.html) * [R crash course](https://github.com/ColauttiLab/RCrashCourse_Book)

**Python**: [Conda on O2](python/conda.md)

**[Single Cell RNA-seq](scrnaseq/README.md):** [Bcbio indrops3](https://bcbio-nextgen.readthedocs.io/en/latest/contents/single_cell.html#workflow) * [Rstudio docker](scrnaseq/rstudio_sc_docker.md) * [Saturation](scrnaseq/saturation_qc.md) * [Clustering analysis in Seurat](scrnaseq/seurat_clustering_analysis.md) * [Seurat markers](scrnaseq/seurat_markers.md) * [Single Cell conda](scrnaseq/Single-Cell-conda.md) * [Cell types](scrnaseq/tinyatlas.md) * [Tools](https://github.com/seandavi/awesome-single-cell) * [Tutorials](scrnaseq/tutorials.md) * [Bibliography](scrnaseq/bibliography.md) * [Velocity](scrnaseq/velocity.md) * [Doublets](scrnaseq/doublets.md) * [Zinbwaver](scrnaseq/zinbwaver.md) * [pseudobulk DESeq2](https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/pseudobulk_DESeq2_scrnaseq.md) * [pseudobulk edgeR](scrnaseq/pseudobulkDE_edgeR.md) * [CITE-seq](scrnaseq/cite_seq.md) * [Library structures](https://teichlab.github.io/scg_lib_structs/) * [Seurat to 10X counts](scrnaseq/write10Xcounts.md) * [Azimuth](https://azimuth.hubmapconsortium.org/) * [Running Cell Ranger O2](scrnaseq/CellRanger.md) * [Cluster annotation tutorial](https://www.nature.com/articles/s41596-021-00534-0) * [Diff composition analysis](https://github.com/theislab/scCODA) * [OSCA](https://bioconductor.org/books/release/OSCA/) * [upstream TF with pySCENIC](https://github.com/hbc/knowledgebase/blob/master/scrnaseq/pySCENIC.md) * [SC article list for beginners](https://threadreaderapp.com/thread/1577714756047278080.html) * [demuxafy (includes demultiplexing, background and doublet detection tools](https://github.com/hbc/knowledgebase/blob/master/scrnaseq/Demuxafy_HowTo.md) * [Lymphocyte Antigen Receptor Transcripts and Clustering](https://github.com/hbc/knowledgebase/blob/master/scrnaseq/Thoughts_on_lymphocyte_Antigen_Receptor_transcripts_in_sc_RNA_Seq_analyses.md) * [DoubletFinder](https://github.com/hbc/knowledgebase/blob/master/scrnaseq/running_doubletfinder.md) * [Running Mast](https://github.com/hbc/knowledgebase/blob/master/scrnaseq/running_MAST.md) * [Publication Quality UMAPs](https://github.com/hbc/knowledgebase/blob/master/scrnaseq/pub_quality_umaps.md)

**Spatial Transcriptomics**: [Running Baysor on o2](https://github.com/hbc/knowledgebase/blob/master/SpatialTranscriptomics/Baysor/RunningBaysor.md) * [Baysor Output into a Seurat Object](https://github.com/hbc/knowledgebase/blob/master/SpatialTranscriptomics/Baysor/LoadingBaysorSegmentationIntoSeurat.R)

**Training Team**: [GitHub page](https://github.com/hbctraining) * [Training Webpage](http://bioinformatics.sph.harvard.edu/training) * [Workshop materials](https://hbctraining.github.io/main) * [HTML reference for GitHub-flavored Markdown](https://hbctraining.github.io/main/GitHub_HTML_reference.html) * [GitHub Pages Documentation](https://hbctraining.github.io/main/GitHub_pages/Pages_documentation.html)
- Team members: Meeta Mistry, Will Gammerdinger, Heather Wick, Elizabeth Partan, Noor Sohail, Upen Bhattarai
- Email: [hbctraining (at) hsph.harvard.edu](mailto:hbctraining@hsph.harvard.edu)

**Variants**: [Clonal evolution](variants/clonal_evolution.md) * [Cancer variant databases](https://github.com/seandavi/awesome-cancer-variant-databases) * [Open Cravat](https://open-cravat.readthedocs.io/en/latest/index.html#) * [Applied computational genomics course by A. Quinlan](https://github.com/quinlan-lab/applied-computational-genomics) *[SC Phylogeny](https://github.com/amkozlov/cellphy) * [Validating variants in IGV](https://bioinformatics-core-shared-training.github.io/intro-to-IGV/InspectingVariantsInIGV.html) * [Human Genome Variation Lab](https://mccoy-lab.github.io/hgv_modules/index.html)

**WGS**: [CRISPR offtarget](wgs/crispr-offtarget.md) * [Pacbio genome assembly](wgs/pacbio_genome_assembly.md)

**WGBS**: [BCratio using centromere as spike-in](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-3334-z) * [RnBeads](https://www.bioconductor.org/packages/release/bioc/html/RnBeads.html)

***

# How it is organized
- minimal overhead: feel free to add any piece of knowledge in any way
- easy to navigate in one click
- github is tracking changes for you
- it is ok to have multiple similar recipes
- growing naturally
- no placeholders
- please name all pages in small_letters.md
- order matters: topics come alpabetically, but within topic and within page please put important things first (important articles/tools on top, important links on the left).

# How to contribute
- nagivate to where you want to share your knowledge or fix something
- don't create a new folder unless you have >=2 new pages that don't fit into existing structure
- add actual value (know how), not structure
- hit 'edit file' or 'new file' (add **.md** extension)
- translate knowledge from your mind into markdown
- hit 'preview changes' to make sure your formatting is fine
- hit 'commit changes'
- thank you, you just have saved somebody's day!

# Knowledge we need to extract (aka TODO):
- ATAC-seq
- miRNA
- integrate (link?) hbctraining, tutorials, bcbio docs
- https://github.com/hbc/hbcABC
- https://github.com/vbarrera/bcbioLite
- Lorena's wisdom:
  - information that apply to all templates: http://bioinformatics.sph.harvard.edu/hbcABC/articles/general_start.html
  - List of templates right now: http://bioinformatics.sph.harvard.edu/hbcABC/articles/list_of_templates.html
  - single cell custom templates: http://bioinformatics.sph.harvard.edu/hbcABC/articles/singlecell-tutorial.html
- Victor's wisdom:
  - Containerization and automated setup of Docker or Conda envs for a project
  - Singularity on O2
  - Running nf-core pipelines on O2 

